packages/core/src/QueryBuilder/Rule/Value/Value.tsx

```diff
@@ -19,13 +21,56 @@ export const Value = ({
   operator,
   value: valueProp,
 }: ValueProps) => {
-  const { attributes, initialTouched, renderers } = useQueryBuilderContext();
+  const { attributes, initialTouched, renderers, emptyRenderer } =
+    useQueryBuilderContext();
 
-  const value =
-    attribute && attributes ? { ...attributes[attribute] } : { type: null };
-  const { type } = value;
+  const attrType =
+    attribute && attributes ? attributes[attribute].type : undefined;
 
-  switch (type) {
+  // Empty value renderer
+  if (emptyRenderer?.find((op) => op === operator)) {
+    return <EmptyValue id={id} />;
+  }
+
+  // Custom renderer
+  if (attrType && renderers?.[attrType]) {
+    // Renderer to be used by order:
+    // 1. Custom attribute renderer
+    // 2. Custom attribute-operator renderer
+    // 3. Custom attribute-DEFAULT renderer
+    // 4. Custom master DEFAULT renderer
+    // 5. Custom master DEFAULT-operator renderer
+    // 6. Custom master DEFAULT-DEFAULT renderer
+    const Renderer: React.FC<HvQueryBuilderRendererProps> | undefined =
+      (typeof renderers[attrType] === "function" && renderers[attrType]) ||
+      (typeof renderers[attrType] === "object" &&
+        operator &&
+        renderers[attrType][operator]) ||
+      (typeof renderers[attrType] === "object" &&
+        renderers[attrType][defaultRendererKey]) ||
+      (typeof renderers[defaultRendererKey] === "function" &&
+        renderers[defaultRendererKey]) ||
+      (typeof renderers[defaultRendererKey] === "object" &&
+        operator &&
+        renderers[defaultRendererKey][operator]) ||
+      (typeof renderers[defaultRendererKey] === "object" &&
+        renderers[defaultRendererKey][defaultRendererKey]) ||
```

Comment:
There's 2 big similar "chains" here - checking for `renderers[attrType]` and then `renderers[defaultRendererKey]`. Maybe a refactor like so would make it easier to follow

```suggestion
      getRenderer(renderers[attrType], operator) ||
      getRenderer(renderers[defaultRendererKey], operator) ||
      undefined;
```

Reply:
```tsx
const getRenderer = (renderer: HvQueryBuilderRenderers, operator: string) =>
  // 1. Custom renderer
  (typeof renderer === "function" && renderer) ||
  // 2. Custom operator renderer
  (typeof renderer === "object" && operator && renderer[operator]) ||
  // 3. Custom DEFAULT renderer
  (typeof renderer === "object" && renderer[defaultRendererKey]);
```

Reply:
Good call, I'll add this 🙏